### PR TITLE
fix(core): CfnMapping values cannot be used in other stacks

### LINKED ISSUE #18920: (core): CfnMapping in one stack, use value in another
### What is the problem?

CfnMapping in one stack, use value in another.

Will lead to "unknown mapping" error in 2nd stack.

### PR DESCRIPTION
**Pull request by comcalvi**
Creating a `CfnMapping` in one stack and then using `findInMap()` in another stack will result in an error at deploy time because the `findInMap()` will turn into a `Fn::FindInMap` that searches for a mapping that does not exist. This change copies the mapping to any stack that references it, even if it is not defined.

Closes #18920.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by comcalvi** [comment]
Certain characters (like `/` or `--`) are not being correctly transformed, causing the ID check to incorrectly think an ID is missing.

File: packages/@aws-cdk/core/lib/cfn-mapping.ts [hunk 1]
```diff
@@ -41,7 +43,7 @@ export class CfnMapping extends CfnRefElement {
   private lazyRender = false;
   private lazyInformed = false;
 
-  constructor(scope: Construct, id: string, props: CfnMappingProps = {}) {
+  constructor(scope: Construct, readonly id: string, props: CfnMappingProps = {}) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/cfn-mapping.ts:46]: `this.node.id` already stores this value.

File: packages/@aws-cdk/core/lib/cfn-mapping.ts [hunk 2]
```diff
@@ -123,3 +126,41 @@ export class CfnMapping extends CfnRefElement {
...
+    }
+
+    return false;
+  }
+
+  public resolve(context: IResolveContext): string {
+    const consumingStack = Stack.of(context.scope);
+    // we must check the entire tree because CfnMappings can be
+    // added to Constructs and won't be direct children of the
+    // stack. This is done in aws-secretsmanager/lib/secret-rotation.ts
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/cfn-mapping.ts:153]: We can make it slightly simpler though.

If `Stack.of(context.scope) === Stack.of(cfnMapping)`, we don't need to do anything anyway -- so we don't care where in the tree it is!

And if we copy it to another stack, we can decide where to put it (directly under the Stack would be an obvious place).

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/core/lib/cfn-mapping.ts (hunk 2 above)
Comment (comcalvi) [reply packages/@aws-cdk/core/lib/cfn-mapping.ts:153]: I like the idea, but this doesn't quite work; if `resolve()` is called more than once in that context, this check will not block us from trying to create the mapping again.

**Review by rix0rrr — COMMENTED** [review]
Please make sure that your PR title confirms to the conventional commit standard (`fix`, `feat`, `chore`) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/v1-main/CONTRIBUTING.md#step-4-pull-request))

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**